core/src/main/java/hudson/cli/listeners/CliListener.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2025, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package hudson.cli.listeners;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.ExtensionPoint;
+import hudson.cli.CLICommand;
+
+/**
+ * Allows implementations to listen to {@link CLICommand#run()} execution events.
+ *
+ * @since TODO
+ */
+public interface CliListener extends ExtensionPoint {
+
+    /**
+     * Invoked before command execution.
+     *
+     * @param context Information about the command being executed
+     * */
+    default void onExecution(@NonNull CliContext context) {}
+
+    /**
+     * Invoked after command execution.
+     *
+     * @param context Information about the command being executed
+     * @param exitCode Exit code returned by the implementation of {@link CLICommand#run()}.
+     * */
+    default void onCompleted(@NonNull CliContext context, int exitCode) {}
+
+    /**
+     * Invoked on command exception.
+     *
+     * @param context Information about the command being executed
+     * @param t Any error during the execution of the command.
+     * */
+    default void onException(@NonNull CliContext context, @NonNull Throwable t) {}
```

Comment:
Misleading name if it takes a `Throwable`? `#onError` maybe?

Reply:
I renamed `onError` to `onException` to better emphasize that `onComplete` also handles cases where `run` returns an error code.
https://github.com/jenkinsci/jenkins/pull/10382#discussion_r1989712724
I'm open to suggestions, but I prefer to avoid using `onError`.

Reply:
JUnit has failures (unsatisfied assertions) and errors (unexpected exceptions thrown), perhaps we could use one term to refer to "did not succeed" and use the other for "this threw up"?